### DISCUSSION TITLE
fenced_code doesn't play well with attr_list

### DISCUSSION THREAD

**Discussion by yengip** [django-wiki/django-wiki#1335] [Q&A]
Hi,

Am I right in thinking that the replacement for markdown.extensions.fenced_code that is embedded in core.markdown.mdx.codehilite will prevent the use of attr_list entries for fenced code blocks?  It looks like it's only possible to specify a language for those blocks (plus something called highlight lines although I can't work out what they're for or how they might work).

I was looking for a way to override the default highlighting behaviour and I wanted to include `{ use_pygment=false }` as suggested in the [docs](https://python-markdown.github.io/extensions/fenced_code_blocks/#enabling-syntax-highlighting), but I couldn't get it to work, and I think this is why.

For now, I've included a config parameter in my project settings like: 
```
WIKI_MARKDOWN_KWARGS = {
    ...
    'extension_configs': {'wiki.plugins.macros.mdx.toc': {'title': 'Contents'},
                          'wiki.core.markdown.mdx.codehilite': {'guess_lang': False}
                         }
```
That won't apply codehilite unless the language is specified using e.g. `{.python}`.  So I get what I want for this specific use case although it was a bit of a trial to work out.

But, I also want to try and apply non-scrolling class(es) to certain code blocks so I can display more than 5-6 lines without scrolling, and maybe extend the default width a bit, but without attr_list I think I'm stuck embedding `<style>` tags in my markdown.  Is there a better way?

Thanks,

Ian

**Comment by benjaoming** [comment]
Hi Ian :wave: 

Since django-wiki is a Django app, you essentially setup a Django project to make it work. Django projects have the ability to overwrite templates. You can read more about it here: https://django-wiki.readthedocs.io/en/main/customization.html#templates

So what I suggest you to do is to add your own `base.html` or `base_site.html` and then inject a `<style>` tag somewhere appropriate. You'll have to look around at the existing templates to understand how this works. Note that you can use `{{ block.super }}` to avoid overwriting everything, see: https://docs.djangoproject.com/en/5.0/ref/templates/language/#template-inheritance

**Comment by yengip** [comment]
Hi Benjamin,

Thanks for the rapid response.  I have dabbled with overriding templates
but really want to be able to control this behaviour in the article rather
than as a blanket thing that effects all articles and all fenced code
blocks.  Which was why I was looking at being able to attach a class to a
fenced code block so it could use my custom class from the overridden
template.

IP

On Thu, 1 Feb 2024 at 07:49, Benjamin Balder Bach ***@***.***>
wrote:

> Hi Ian 👋
>
> Since django-wiki is a Django app, you essentially setup a Django project
> to make it work. Django projects have the ability to overwrite templates.
> You can read more about it here:
> https://django-wiki.readthedocs.io/en/main/customization.html#templates
>
> So what I suggest you to do is to add your own base.html or base_site.html
> and then inject a <style> tag somewhere appropriate. You'll have to look
> around at the existing templates to understand how this works. Note that
> you can use {{ block.super }} to avoid overwriting everything, see:
> https://docs.djangoproject.com/en/5.0/ref/templates/language/#template-inheritance
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/django-wiki/django-wiki/discussions/1335#discussioncomment-8331635>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUWY2KFMMF7DLZFUVB6D3ZTYRNCIHAVCNFSM6AAAAABCUJBJGOVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DGMZRGYZTK>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>


-- 
Best regards,

Ian Pigney

M: +44 7967 381 829

**Reply by benjaoming** [reply]
Aaah, I see what you mean with this attr_list: https://python-markdown.github.io/extensions/fenced_code_blocks/#attributes

The implementation is here:

https://github.com/django-wiki/django-wiki/blob/a64a4f4055010ec25b916d6b3121dee22749cba7/src/wiki/core/markdown/mdx/codehilite.py#L37-L47

It may not be implementing the latest behavior.

If you are able to find an error or improve this for your usecase, you are more than welcome to say so or open a PR.

**Reply by yengip** [reply]
I think perhaps I've misunderstood how the various extensions interact with each other and the text.  I'll need to do a lot more research and get a proper grasp of that before I can work out what's happening and understand why the current implementation doesn't seem to allow attributes or extra classes in fenced code blocks.  The docs say they should work together but they're not.
I'll stick with my settings workaround for now until I can find enough spare time to get studying.
Thanks.

**Reply by benjaoming** [reply]
I'm sorry that I never answered your initial question:

> Am I right in thinking that the replacement for markdown.extensions.fenced_code that is embedded in core.markdown.mdx.codehilite will prevent the use of attr_list entries for fenced code blocks?

I don't know the expected result of adding two Markdown extensions that consume the same syntax. But I think that it's likely gonna mean that one of them takes precedence over the next. For django-wiki, what you want is probably to fix `wiki.core.markdown.mdx.codehilite` to parse and understand `attr_list`. This needs to be fixed since the django-wiki extension is a spin-off of an older version of `fenced_code`.

**Reply by yengip** [reply]
No worries.  I'm keen to contribute but I am struggling to find time to set up a clone repo and educate myself on how the whole extensions thing works right now.  I'll get round to it ;-)

**Reply by benjaoming** [reply]
I've opened an issue from this discussion: https://github.com/django-wiki/django-wiki/issues/1356
